src/i2c.rs

```diff
@@ -197,26 +322,100 @@ impl<'a, Device: FtdiCommon> embedded_hal::blocking::i2c::Write for I2c<'a, Devi
         let mut ack_buf: Vec<u8> = vec![0; 1 + bytes.len()];
-        inner.ft.read_all(ack_buf.as_mut_slice())
+        inner.ft.read_all(ack_buf.as_mut_slice())?;
+        if ack_buf.iter().any(|&ack| (ack & 0b1) != 0x00) {
+            Err(I2cError::Nak)
+        } else {
+            Ok(())
+        }
     }
-}
 
-impl<'a, Device: FtdiCommon> embedded_hal::blocking::i2c::WriteRead for I2c<'a, Device> {
-    type Error = TimeoutError;
+    fn write_slow(&mut self, addr: u8, bytes: &[u8]) -> Result<(), I2cError> {
+        assert!(!bytes.is_empty(), "bytes must be a non-empty slice");
 
-    fn write_read(
+        let lock = self.mtx.lock().expect("Failed to aquire FTDI mutex");
+        let mut inner = lock.borrow_mut();
+
+        // ST
+        let mut mpsse_cmd: MpsseCmdBuilder = MpsseCmdBuilder::new();
+        for _ in 0..self.start_stop_cmds {
+            mpsse_cmd =
+                mpsse_cmd.set_gpio_lower(SCL | SDA | inner.value, SCL | SDA | inner.direction)
+        }
+        for _ in 0..self.start_stop_cmds {
+            mpsse_cmd = mpsse_cmd.set_gpio_lower(SCL | inner.value, SCL | SDA | inner.direction)
+        }
+
+        mpsse_cmd = mpsse_cmd
+            // SAD+W
+            .set_gpio_lower(inner.value, SCL | SDA | inner.direction)
+            .clock_bits_out(BITS_OUT, addr << 1, 8)
+            // SAK
+            .set_gpio_lower(inner.value, SCL | inner.direction)
+            .clock_bits_in(BITS_IN, 1)
+            .send_immediate();
+
+        inner.ft.write_all(mpsse_cmd.as_slice())?;
+        let mut ack_buf: [u8; 1] = [0; 1];
+        inner.ft.read_all(&mut ack_buf)?;
+        if (ack_buf[0] & 0b1) != 0x00 {
+            return Err(I2cError::Nak);
+        }
+
+        for (idx, byte) in bytes.iter().enumerate() {
+            let mut mpsse_cmd: MpsseCmdBuilder = MpsseCmdBuilder::new()
+                // Bi
+                .set_gpio_lower(inner.value, SCL | SDA | inner.direction)
+                .clock_bits_out(BITS_OUT, *byte, 8)
+                // SAK
+                .set_gpio_lower(inner.value, SCL | inner.direction)
+                .clock_bits_in(BITS_IN, 1);
+
+            // last byte
+            if idx == bytes.len() - 1 {
+                // SP
+                for _ in 0..self.start_stop_cmds {
+                    mpsse_cmd = mpsse_cmd.set_gpio_lower(inner.value, SCL | SDA | inner.direction)
+                }
+                for _ in 0..self.start_stop_cmds {
+                    mpsse_cmd =
+                        mpsse_cmd.set_gpio_lower(inner.value | SCL, SCL | SDA | inner.direction)
+                }
+                for _ in 0..self.start_stop_cmds {
+                    mpsse_cmd = mpsse_cmd
+                        .set_gpio_lower(inner.value | SCL | SDA, SCL | SDA | inner.direction)
+                }
+            }
+
+            mpsse_cmd = mpsse_cmd
+                // Idle
+                .set_gpio_lower(inner.value, inner.direction)
+                .send_immediate();
```

Comment:
You have to move this up into the last byte condition. Currently it fails when you write more than one byte using ``write_slow``.

Reply:
Good catch!  Made the change.